### PR TITLE
Refactor Tactic FSMs to Move Function Implementations to Source Files

### PR DESCRIPTION
**Pull request by jonathanlew**
### Description
High level idea is to prepare for the new play architecture with better code quality. More details:
* Create macros for defining lambda wrappers around state, event, action, and guard
* Remove constructor arguments where not needed
* Move all function implementations to *_fsm.cpp files
* Rename DEFINE_UPDATE_STRUCT_WITH_CONTROL_AND_COMMON_PARAMS to DEFINE_TACTIC_UPDATE_STRUCT_WITH_CONTROL_AND_COMMON_PARAMS
* Move common SML definitions to util folder

### Testing Done
regression tests pass

### Length Justification
:( It's refactoring so there shouldn't be any functional changes

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Review by akhilveeraghanta — DISMISSED** [review]
noice! I didn't check every single file, but really liking this improvement

transitional table is way more readable

File: src/software/ai/hl/stp/tactic/stop/stop_fsm.h [hunk 1]
```diff
@@ -1,5 +1,7 @@
 #pragma once
 
+#include <functional>
```
Comment (itsarune) [inline src/software/ai/hl/stp/tactic/stop/stop_fsm.h]: what's this for?

File: src/software/util/sml_fsm/sml_fsm.h [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+/**
+ * Defines an sml state wrapper around a class/struct
+ *
+ * @param STATE The state class/struct
+ */
+#define DEFINE_SML_STATE(STATE) const auto STATE##_S = boost::sml::state<STATE>;
+
+/**
+ * Defines an sml event wrapper around a class/struct
```
Comment (itsarune) [inline src/software/util/sml_fsm/sml_fsm.h]: nit: "sml" and "SML" are being used interchangeably in this file's comments, probably just use one or the other

**Review by itsarune — COMMENTED** [review]
🔥🔥 a nit or two

**Review by akhilveeraghanta — APPROVED** [review]

File: src/software/util/sml_fsm/sml_fsm.h (hunk 2 above)
Comment (jonathanlew) [reply src/software/util/sml_fsm/sml_fsm.h]: fixed

File: src/software/ai/hl/stp/tactic/stop/stop_fsm.h (hunk 1 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/stop/stop_fsm.h]: fixed

**Review by itsarune — APPROVED** [review]

**Pull request merged**